Api/Controllers/ProductsController.cs

```diff
@@ -0,0 +1,194 @@
+
+            var models = dTOs.Select(ProductModel.MapFrom).ToList();
+
+            foreach (var product in models)
+            {
+                var links = CreateLinksForProduct(categoryId, product.Id);
+                product.Links = links.ToList();
+            }
+
+
+            var productsWrapper = new LinkCollectionWrapper<ProductModel>(models);
+            return Ok(CreateLinksForGetAllProducts(productsWrapper));
+        }
+
+        // GET api/<ProductsController>/5
+        [HttpGet("{id}")]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProductModel), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> GetById(int categoryId, int id)
+        {
+            var productDto = await _productService.GetProductByIdAsync(id);
+            var categoryDto = await _categoryService.GetCategoryByIdAsync(categoryId);
+
+            if (productDto == null || categoryDto == null)
+                return NotFound();
+
+            var model = ProductModel.MapFrom(productDto);
+            model.Links = CreateLinksForProduct(categoryId, model.Id).ToList();
+
+            return Ok(model);
+        }
+
+        // POST api/<ProductsController>
+        [HttpPost]
+        [Consumes("application/json")]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProductModel), StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<IActionResult> Post(int categoryId, [FromBody] ProductWriteModel model)
+        {
+            var categoryDto = await _categoryService.GetCategoryByIdAsync(categoryId);
+
+            if (categoryDto == null)
+                return NotFound();
+
+            var dto = ProductWriteModel.MapTo(model);
+            dto.CategoryId = categoryId;
+
+            ProductDTO? addedDto = null;
+
+            try
+            {
+                addedDto = await _productService.AddProductAsync(dto);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, $"Add product error");
+                throw;
+            }
+
+            var addedModel = ProductModel.MapFrom(addedDto);
+            addedModel.Links = CreateLinksForProduct(categoryId, addedModel.Id).ToList();
+
+            return Created(nameof(GetById), addedModel);
+        }
+
+        // PUT api/<ProductsController>/5
+        [HttpPut("{id}")]
+        [Consumes("application/json")]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProductModel), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<IActionResult> Update(int categoryId, int id, [FromBody] ProductWriteModel model)
+        {
```

Comment:
if this endpoint was accessible by `PUT /api/products/{id}` route, it would make it much simpler. Consider doing this

Reply:
I strongly suggest to refactor this controller and exclude categories from the route.
For a bit more information, check top answer here: https://stackoverflow.com/questions/20951419/what-are-best-practices-for-rest-nested-resources

It seems to me that even in this simple API nested resources introduce more problems than they solve